logr.go

```diff
@@ -382,12 +477,90 @@ func (l Logger) IsZero() bool {
 	return l.sink == nil
 }
 
+// WithContext stores a context in the Logger. If the Logger also has keys set
+// that it is meant to log from a context, then the values for those keys from
+// the given context will be added to all log entries.
+func (l Logger) WithContext(ctx context.Context) Logger {
+	if l.sink == discardLogSinkInstance {
+		return l
+	}
+
+	l.setContext(ctx)
+	return l
+}
+
+// WithContextHandlers extends the list of context handlers. When
+// given a context through WithContext later, the Logger will invoke
+// each handler and log the key/value pairs that they return as
+// as if they had been passed to WithValues.
+func (l Logger) WithContextHandlers(handlers ...ContextHandler) Logger {
+	if l.sink == discardLogSinkInstance {
+		return l
+	}
+
+	l.setContextHandlers(handlers...)
+	return l
+}
+
+// mergeHandlers combines different slices into one. If a single one is passed,
+// it will still create a new slice and copy handlers to make the result
+// independent.
+func mergeHandlers(handlerSlices ...[]ContextHandler) []ContextHandler {
+	n := 0
+	for _, handlers := range handlerSlices {
+		n += len(handlers)
+	}
+	if n == 0 {
+		return nil
+	}
+	cp := make([]ContextHandler, 0, n)
+	for _, handlers := range handlerSlices {
+		cp = append(cp, handlers...)
+	}
+	return cp
+}
+
+func (l Logger) valuesFromContext() []interface{} {
+	if l.contextKV == nil || l.contextKV.ctx == nil || len(l.contextKV.handlers) == 0 {
+		return nil
+	}
+
+	// Fast path for a single handler: just return its result directly.
+	if len(l.contextKV.handlers) == 1 {
+		return l.contextKV.handlers[0].ValuesFromContext(l.contextKV.ctx)
+	}
+
+	// We don't really know how large this has to be. Let's
+	// assume that each handler returns at most two key/value
+	// pairs, of which each needs two entries.
+	keysAndValues := make([]interface{}, 0, 64*len(l.contextKV.handlers))
+	for _, handler := range l.contextKV.handlers {
+		keysAndValues = append(keysAndValues, handler.ValuesFromContext(l.contextKV.ctx)...)
+	}
+	return keysAndValues
+}
+
+// ContextHandler provides callbacks for working with a context.
+// More functionality might get added in the future.
+type ContextHandler struct {
+	// ValuesFromContext returns additional key/value pairs for data that
+	// was stored in the context. These pairs will be logged via WithValues
+	// and thus may get overridden by explicit log parameters if there is a
+	// key conflict.
+	//
+	// ValuesFromContext must always return the same key/value pairs for
+	// the same context. The result might get cached.
+	ValuesFromContext func(ctx context.Context) []interface{}
```

Comment:
One more idea: with `ValuesFromContext func(ctx context.Context, buffer []interface{}) []interface{}` the caller can pass in a buffer that was allocated on the stack:
```
kvBuffer := [32]interface{}
kvs := ValuesFromContext(ctx, kvBuffer[0:0])
```

The handler can `append` to that buffer and then return the resulting slice. If it appends <= 16 key/value pairs, then no allocation should be necessary.

I still need to try this, but I think it'll work.

Reply:
This is surprising. The amount of allocations remained the same, but their size increased and the tests became slower?! I don't understand.

Here's the patch:
```diff
diff --git a/discard_test.go b/discard_test.go
index db4c4db..143afd3 100644
--- a/discard_test.go
+++ b/discard_test.go
@@ -48,7 +48,7 @@ func TestComparable(t *testing.T) {
 		t.Fatal("any two discardLogSink must be equal")
 	}
 
-	if b := a.WithContextHandlers(ContextHandler{func(ctx context.Context) []interface{} { return nil }}); b != a {
+	if b := a.WithContextHandlers(ContextHandler{func(ctx context.Context, buffer []interface{}) []interface{} { return buffer }}); b != a {
 		t.Fatalf("Discard logger should not get modified by WithContextValues, but it was: %+v", b)
 	}
 
diff --git a/examples/benchmarks/contextvalues_test.go b/examples/benchmarks/contextvalues_test.go
index 8a6daec..ebba723 100644
--- a/examples/benchmarks/contextvalues_test.go
+++ b/examples/benchmarks/contextvalues_test.go
@@ -156,14 +156,14 @@ func BenchmarkNewContextMany(b *testing.B) {
 type contextKey1 struct{}
 type contextKey2 struct{}
 
-func fromContext(ctx context.Context) []interface{} {
+func fromContext(ctx context.Context, buffer []interface{}) []interface{} {
 	value1 := ctx.Value(contextKey1{})
 	value2 := ctx.Value(contextKey2{})
 	if value1 == nil || value2 == nil {
 		// Should never happen during these tests.
 		panic("failed to retrieve both values")
 	}
-	return []interface{}{"i", value1, "j", value2}
+	return append(buffer, "i", value1, "j", value2)
 }
 
 func run(ctx context.Context, op func(ctx context.Context)) {
diff --git a/logr.go b/logr.go
index 35286bd..f244bc7 100644
--- a/logr.go
+++ b/logr.go
@@ -362,7 +362,10 @@ func (l Logger) sinkWithContextValues() LogSink {
 		return value.(LogSink) // nolint: forcetypeassert
 	}
 	sink := l.sink
-	contextKeysAndValues := l.valuesFromContext()
+	// This array on the stack is large enough for 32 key/value pairs.
+	// The slice initially has that capacity and zero entries.
+	var buffer [64]interface{}
+	contextKeysAndValues := l.valuesFromContext(buffer[0:0])
 	if len(contextKeysAndValues) > 0 {
 		sink = sink.WithValues(contextKeysAndValues...)
 	}
@@ -504,37 +507,30 @@ func mergeHandlers(handlerSlices ...[]ContextHandler) []ContextHandler {
 	return cp
 }
 
-func (l Logger) valuesFromContext() []interface{} {
+func (l Logger) valuesFromContext(buffer []interface{}) []interface{} {
 	if l.contextKV == nil || l.contextKV.ctx == nil || len(l.contextKV.handlers) == 0 {
-		return nil
-	}
-
-	// Fast path for a single handler: just return its result directly.
-	if len(l.contextKV.handlers) == 1 {
-		return l.contextKV.handlers[0].ValuesFromContext(l.contextKV.ctx)
+		return buffer
 	}
 
-	// We don't really know how large this has to be. Let's
-	// assume that each handler returns at most two key/value
-	// pairs, of which each needs two entries.
-	keysAndValues := make([]interface{}, 0, 64*len(l.contextKV.handlers))
 	for _, handler := range l.contextKV.handlers {
-		keysAndValues = append(keysAndValues, handler.ValuesFromContext(l.contextKV.ctx)...)
+		buffer = handler.ValuesFromContext(l.contextKV.ctx, buffer)
 	}
-	return keysAndValues
+	return buffer
 }
 
 // ContextHandler provides callbacks for working with a context.
 // More functionality might get added in the future.
 type ContextHandler struct {
-	// ValuesFromContext returns additional key/value pairs for data that
-	// was stored in the context. These pairs will be logged via WithValues
-	// and thus may get overridden by explicit log parameters if there is a
-	// key conflict.
+	// ValuesFromContext stores additional key/value pairs in the buffer
+	// for data that was stored in the context. It returns the extended
+	// slice.
+	//
+	// These pairs will be logged via WithValues and thus may get
+	// overridden by explicit log parameters if there is a key conflict.
 	//
-	// ValuesFromContext must always return the same key/value pairs for
+	// ValuesFromContext must always add the same key/value pairs for
 	// the same context. The result might get cached.
-	ValuesFromContext func(ctx context.Context) []interface{}
+	ValuesFromContext func(ctx context.Context, buffer []interface{}) []interface{}
 }
 
 // contextKey is how we find Loggers in a context.Context.

```


And the benchmark.

<details>

````
name                                                     old time/op    new time/op    delta
pkg:github.com/go-logr/logr/benchmark goos:linux goarch:amd64
DiscardLogInfoOneArg-36                                    96.5ns ± 4%    99.7ns ± 2%     ~     (p=0.095 n=5+5)
DiscardLogInfoSeveralArgs-36                                239ns ± 1%     239ns ± 1%     ~     (p=0.524 n=5+5)
DiscardLogInfoWithValues-36                                 240ns ± 1%     243ns ± 1%   +1.23%  (p=0.040 n=5+5)
DiscardLogV0Info-36                                         240ns ± 1%     241ns ± 1%     ~     (p=0.167 n=5+5)
DiscardLogV9Info-36                                         241ns ± 1%     240ns ± 1%     ~     (p=0.341 n=5+5)
DiscardLogError-36                                          248ns ± 0%     247ns ± 1%     ~     (p=0.690 n=5+5)
DiscardWithValues-36                                        122ns ± 1%     119ns ± 1%   -2.12%  (p=0.016 n=5+5)
DiscardWithName-36                                         3.80ns ± 0%    3.79ns ± 0%     ~     (p=0.198 n=5+5)
FuncrLogInfoOneArg-36                                      2.10µs ± 2%    2.14µs ± 2%     ~     (p=0.095 n=5+5)
FuncrJSONLogInfoOneArg-36                                  2.39µs ± 1%    2.39µs ± 2%     ~     (p=1.000 n=5+5)
FuncrLogInfoSeveralArgs-36                                 4.49µs ± 3%    4.52µs ± 1%     ~     (p=1.000 n=5+5)
FuncrJSONLogInfoSeveralArgs-36                             4.76µs ± 2%    4.80µs ± 6%     ~     (p=0.151 n=5+5)
FuncrLogInfoWithValues-36                                  4.56µs ± 2%    4.55µs ± 2%     ~     (p=0.690 n=5+5)
FuncrJSONLogInfoWithValues-36                              4.83µs ± 3%    4.83µs ± 2%     ~     (p=1.000 n=5+5)
FuncrLogV0Info-36                                          4.42µs ± 5%    4.56µs ± 1%     ~     (p=0.151 n=5+5)
FuncrJSONLogV0Info-36                                      4.73µs ± 5%    4.84µs ± 3%     ~     (p=0.222 n=5+5)
FuncrLogV9Info-36                                           247ns ± 1%     248ns ± 0%     ~     (p=0.413 n=5+5)
FuncrJSONLogV9Info-36                                       245ns ± 1%     249ns ± 1%   +1.46%  (p=0.008 n=5+5)
FuncrLogError-36                                           4.80µs ± 1%    4.85µs ± 3%     ~     (p=0.310 n=5+5)
FuncrJSONLogError-36                                       4.96µs ± 3%    4.94µs ± 1%     ~     (p=0.889 n=5+5)
FuncrWithValues-36                                         1.27µs ± 3%    1.28µs ± 4%     ~     (p=0.841 n=5+5)
FuncrWithName-36                                            180ns ± 1%     180ns ± 1%     ~     (p=1.000 n=5+5)
FuncrWithCallDepth-36                                       166ns ± 1%     170ns ± 0%   +2.61%  (p=0.016 n=5+4)
FuncrJSONLogInfoStringerValue-36                           2.63µs ± 1%    2.60µs ± 2%     ~     (p=0.087 n=5+5)
FuncrJSONLogInfoErrorValue-36                              2.60µs ± 1%    2.60µs ± 0%     ~     (p=0.540 n=5+5)
FuncrJSONLogInfoMarshalerValue-36                          2.61µs ± 1%    2.60µs ± 1%     ~     (p=0.841 n=5+5)
pkg:github.com/go-logr/logr/examples/benchmarks goos:linux goarch:amd64
NewContext1Percent/simple-36                               45.7µs ± 4%    46.7µs ± 1%     ~     (p=0.095 n=5+5)
NewContext1Percent/WithValuesBefore-36                     46.4µs ± 2%    47.7µs ± 2%   +2.79%  (p=0.032 n=5+5)
NewContext1Percent/WithValuesAfter-36                      47.8µs ± 2%    49.5µs ± 2%   +3.53%  (p=0.032 n=5+5)
NewContext1Percent/WithName-36                             46.2µs ± 2%    48.4µs ± 1%   +4.77%  (p=0.008 n=5+5)
NewContext1Percent/WithNameAndValuesAfter-36               48.3µs ± 2%    49.7µs ± 1%   +2.80%  (p=0.032 n=5+5)
NewContext1Percent/WithNameAndValuesBefore-36              47.5µs ± 1%    48.6µs ± 1%   +2.47%  (p=0.008 n=5+5)
NewContext100PercentDisabled/simple-36                     74.0µs ± 2%    73.0µs ± 1%     ~     (p=0.151 n=5+5)
NewContext100PercentDisabled/WithValuesBefore-36           74.4µs ± 2%    73.3µs ± 1%     ~     (p=0.056 n=5+5)
NewContext100PercentDisabled/WithValuesAfter-36             179µs ± 3%     178µs ± 4%     ~     (p=1.000 n=5+5)
NewContext100PercentDisabled/WithName-36                    104µs ± 1%     105µs ± 1%   +1.11%  (p=0.008 n=5+5)
NewContext100PercentDisabled/WithNameAndValuesAfter-36      214µs ± 2%     210µs ± 1%     ~     (p=0.151 n=5+5)
NewContext100PercentDisabled/WithNameAndValuesBefore-36     104µs ± 2%     105µs ± 1%     ~     (p=0.056 n=5+5)
NewContext100Percent/simple-36                              531µs ± 1%     604µs ± 1%  +13.70%  (p=0.008 n=5+5)
NewContext100Percent/WithValuesBefore-36                    603µs ± 2%     681µs ± 1%  +12.91%  (p=0.008 n=5+5)
NewContext100Percent/WithValuesAfter-36                     712µs ± 2%     793µs ± 1%  +11.33%  (p=0.008 n=5+5)
NewContext100Percent/WithName-36                            585µs ± 2%     658µs ± 2%  +12.48%  (p=0.008 n=5+5)
NewContext100Percent/WithNameAndValuesAfter-36              742µs ± 2%     813µs ± 2%   +9.63%  (p=0.008 n=5+5)
NewContext100Percent/WithNameAndValuesBefore-36             623µs ± 2%     709µs ± 1%  +13.93%  (p=0.008 n=5+5)
NewContextMany/simple-36                                   3.69ms ± 1%    3.70ms ± 2%     ~     (p=0.548 n=5+5)
NewContextMany/WithValuesBefore-36                         3.77ms ± 2%    3.81ms ± 2%     ~     (p=0.310 n=5+5)
NewContextMany/WithValuesAfter-36                          3.90ms ± 2%    3.92ms ± 1%     ~     (p=0.690 n=5+5)
NewContextMany/WithName-36                                 3.96ms ± 2%    3.99ms ± 2%     ~     (p=0.095 n=5+5)
NewContextMany/WithNameAndValuesAfter-36                   4.11ms ± 1%    4.16ms ± 1%     ~     (p=0.151 n=5+5)
NewContextMany/WithNameAndValuesBefore-36                  3.98ms ± 1%    4.02ms ± 0%     ~     (p=0.056 n=5+5)

name                                                     old alloc/op   new alloc/op   delta
pkg:github.com/go-logr/logr/benchmark goos:linux goarch:amd64
DiscardLogInfoOneArg-36                                     32.0B ± 0%     32.0B ± 0%     ~     (all equal)
DiscardLogInfoSeveralArgs-36                                 176B ± 0%      176B ± 0%     ~     (all equal)
DiscardLogInfoWithValues-36                                  176B ± 0%      176B ± 0%     ~     (all equal)
DiscardLogV0Info-36                                          176B ± 0%      176B ± 0%     ~     (all equal)
DiscardLogV9Info-36                                          176B ± 0%      176B ± 0%     ~     (all equal)
DiscardLogError-36                                           176B ± 0%      176B ± 0%     ~     (all equal)
DiscardWithValues-36                                        64.0B ± 0%     64.0B ± 0%     ~     (all equal)
DiscardWithName-36                                          0.00B          0.00B          ~     (all equal)
FuncrLogInfoOneArg-36                                      1.15kB ± 0%    1.15kB ± 0%     ~     (all equal)
FuncrJSONLogInfoOneArg-36                                  1.17kB ± 0%    1.17kB ± 0%     ~     (all equal)
FuncrLogInfoSeveralArgs-36                                 1.45kB ± 0%    1.45kB ± 0%     ~     (all equal)
FuncrJSONLogInfoSeveralArgs-36                             1.45kB ± 0%    1.45kB ± 0%     ~     (all equal)
FuncrLogInfoWithValues-36                                  1.48kB ± 0%    1.48kB ± 0%     ~     (all equal)
FuncrJSONLogInfoWithValues-36                              1.50kB ± 0%    1.50kB ± 0%     ~     (all equal)
FuncrLogV0Info-36                                          1.45kB ± 0%    1.45kB ± 0%     ~     (all equal)
FuncrJSONLogV0Info-36                                      1.45kB ± 0%    1.45kB ± 0%     ~     (all equal)
FuncrLogV9Info-36                                            176B ± 0%      176B ± 0%     ~     (all equal)
FuncrJSONLogV9Info-36                                        176B ± 0%      176B ± 0%     ~     (all equal)
FuncrLogError-36                                           1.48kB ± 0%    1.48kB ± 0%     ~     (all equal)
FuncrJSONLogError-36                                       1.50kB ± 0%    1.50kB ± 0%     ~     (all equal)
FuncrWithValues-36                                           264B ± 0%      264B ± 0%     ~     (all equal)
FuncrWithName-36                                            96.0B ± 0%     96.0B ± 0%     ~     (all equal)
FuncrWithCallDepth-36                                       96.0B ± 0%     96.0B ± 0%     ~     (all equal)
FuncrJSONLogInfoStringerValue-36                           1.20kB ± 0%    1.20kB ± 0%     ~     (all equal)
FuncrJSONLogInfoErrorValue-36                              1.20kB ± 0%    1.20kB ± 0%     ~     (all equal)
FuncrJSONLogInfoMarshalerValue-36                          1.20kB ± 0%    1.20kB ± 0%     ~     (all equal)
pkg:github.com/go-logr/logr/examples/benchmarks goos:linux goarch:amd64
NewContext1Percent/simple-36                               19.2kB ± 0%    20.2kB ± 0%   +5.00%  (p=0.008 n=5+5)
NewContext1Percent/WithValuesBefore-36                     19.3kB ± 0%    20.3kB ± 0%   +4.98%  (p=0.008 n=5+5)
NewContext1Percent/WithValuesAfter-36                      19.5kB ± 0%    20.5kB ± 0%   +4.92%  (p=0.008 n=5+5)
NewContext1Percent/WithName-36                             19.4kB ± 0%    20.4kB ± 0%   +4.94%  (p=0.008 n=5+5)
NewContext1Percent/WithNameAndValuesAfter-36               19.7kB ± 0%    20.7kB ± 0%   +4.87%  (p=0.008 n=5+5)
NewContext1Percent/WithNameAndValuesBefore-36              19.5kB ± 0%    20.5kB ± 0%   +4.92%  (p=0.008 n=5+5)
NewContext100PercentDisabled/simple-36                     33.6kB ± 0%    33.6kB ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithValuesBefore-36           33.6kB ± 0%    33.6kB ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithValuesAfter-36            56.0kB ± 0%    56.0kB ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithName-36                   49.6kB ± 0%    49.6kB ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithNameAndValuesAfter-36     72.0kB ± 0%    72.0kB ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithNameAndValuesBefore-36    49.6kB ± 0%    49.6kB ± 0%     ~     (all equal)
NewContext100Percent/simple-36                              178kB ± 0%     274kB ± 0%  +53.81%  (p=0.008 n=5+5)
NewContext100Percent/WithValuesBefore-36                    187kB ± 0%     283kB ± 0%  +51.28%  (p=0.008 n=5+5)
NewContext100Percent/WithValuesAfter-36                     210kB ± 0%     306kB ± 0%  +45.80%  (p=0.008 n=5+5)
NewContext100Percent/WithName-36                            199kB ± 0%     295kB ± 0%  +48.19%  (p=0.016 n=5+4)
NewContext100Percent/WithNameAndValuesAfter-36              230kB ± 0%     326kB ± 0%  +41.81%  (p=0.008 n=5+5)
NewContext100Percent/WithNameAndValuesBefore-36             207kB ± 0%     303kB ± 0%  +46.33%  (p=0.029 n=4+4)
NewContextMany/simple-36                                   1.35MB ± 0%    1.45MB ± 0%   +7.11%  (p=0.008 n=5+5)
NewContextMany/WithValuesBefore-36                         1.37MB ± 0%    1.47MB ± 0%   +6.99%  (p=0.016 n=4+5)
NewContextMany/WithValuesAfter-36                          1.40MB ± 0%    1.49MB ± 0%   +6.88%  (p=0.008 n=5+5)
NewContextMany/WithName-36                                 1.41MB ± 0%    1.51MB ± 0%   +6.79%  (p=0.008 n=5+5)
NewContextMany/WithNameAndValuesAfter-36                   1.46MB ± 0%    1.56MB ± 0%   +6.58%  (p=0.008 n=5+5)
NewContextMany/WithNameAndValuesBefore-36                  1.44MB ± 0%    1.53MB ± 0%   +6.68%  (p=0.016 n=5+4)

name                                                     old allocs/op  new allocs/op  delta
pkg:github.com/go-logr/logr/benchmark goos:linux goarch:amd64
DiscardLogInfoOneArg-36                                      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
DiscardLogInfoSeveralArgs-36                                 2.00 ± 0%      2.00 ± 0%     ~     (all equal)
DiscardLogInfoWithValues-36                                  2.00 ± 0%      2.00 ± 0%     ~     (all equal)
DiscardLogV0Info-36                                          2.00 ± 0%      2.00 ± 0%     ~     (all equal)
DiscardLogV9Info-36                                          2.00 ± 0%      2.00 ± 0%     ~     (all equal)
DiscardLogError-36                                           2.00 ± 0%      2.00 ± 0%     ~     (all equal)
DiscardWithValues-36                                         1.00 ± 0%      1.00 ± 0%     ~     (all equal)
DiscardWithName-36                                           0.00           0.00          ~     (all equal)
FuncrLogInfoOneArg-36                                        7.00 ± 0%      7.00 ± 0%     ~     (all equal)
FuncrJSONLogInfoOneArg-36                                    8.00 ± 0%      8.00 ± 0%     ~     (all equal)
FuncrLogInfoSeveralArgs-36                                   17.0 ± 0%      17.0 ± 0%     ~     (all equal)
FuncrJSONLogInfoSeveralArgs-36                               18.0 ± 0%      18.0 ± 0%     ~     (all equal)
FuncrLogInfoWithValues-36                                    17.0 ± 0%      17.0 ± 0%     ~     (all equal)
FuncrJSONLogInfoWithValues-36                                18.0 ± 0%      18.0 ± 0%     ~     (all equal)
FuncrLogV0Info-36                                            17.0 ± 0%      17.0 ± 0%     ~     (all equal)
FuncrJSONLogV0Info-36                                        18.0 ± 0%      18.0 ± 0%     ~     (all equal)
FuncrLogV9Info-36                                            2.00 ± 0%      2.00 ± 0%     ~     (all equal)
FuncrJSONLogV9Info-36                                        2.00 ± 0%      2.00 ± 0%     ~     (all equal)
FuncrLogError-36                                             19.0 ± 0%      19.0 ± 0%     ~     (all equal)
FuncrJSONLogError-36                                         20.0 ± 0%      20.0 ± 0%     ~     (all equal)
FuncrWithValues-36                                           8.00 ± 0%      8.00 ± 0%     ~     (all equal)
FuncrWithName-36                                             1.00 ± 0%      1.00 ± 0%     ~     (all equal)
FuncrWithCallDepth-36                                        1.00 ± 0%      1.00 ± 0%     ~     (all equal)
FuncrJSONLogInfoStringerValue-36                             10.0 ± 0%      10.0 ± 0%     ~     (all equal)
FuncrJSONLogInfoErrorValue-36                                10.0 ± 0%      10.0 ± 0%     ~     (all equal)
FuncrJSONLogInfoMarshalerValue-36                            10.0 ± 0%      10.0 ± 0%     ~     (all equal)
pkg:github.com/go-logr/logr/examples/benchmarks goos:linux goarch:amd64
NewContext1Percent/simple-36                                  318 ± 0%       318 ± 0%     ~     (all equal)
NewContext1Percent/WithValuesBefore-36                        320 ± 0%       320 ± 0%     ~     (all equal)
NewContext1Percent/WithValuesAfter-36                         326 ± 0%       326 ± 0%     ~     (all equal)
NewContext1Percent/WithName-36                                321 ± 0%       321 ± 0%     ~     (all equal)
NewContext1Percent/WithNameAndValuesAfter-36                  329 ± 0%       329 ± 0%     ~     (all equal)
NewContext1Percent/WithNameAndValuesBefore-36                 323 ± 0%       323 ± 0%     ~     (all equal)
NewContext100PercentDisabled/simple-36                        500 ± 0%       500 ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithValuesBefore-36              500 ± 0%       500 ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithValuesAfter-36             1.10k ± 0%     1.10k ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithName-36                      700 ± 0%       700 ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithNameAndValuesAfter-36      1.30k ± 0%     1.30k ± 0%     ~     (all equal)
NewContext100PercentDisabled/WithNameAndValuesBefore-36       700 ± 0%       700 ± 0%     ~     (all equal)
NewContext100Percent/simple-36                              2.10k ± 0%     2.10k ± 0%     ~     (all equal)
NewContext100Percent/WithValuesBefore-36                    2.30k ± 0%     2.30k ± 0%     ~     (all equal)
NewContext100Percent/WithValuesAfter-36                     2.90k ± 0%     2.90k ± 0%     ~     (all equal)
NewContext100Percent/WithName-36                            2.40k ± 0%     2.40k ± 0%     ~     (all equal)
NewContext100Percent/WithNameAndValuesAfter-36              3.20k ± 0%     3.20k ± 0%     ~     (all equal)
NewContext100Percent/WithNameAndValuesBefore-36             2.60k ± 0%     2.60k ± 0%     ~     (all equal)
NewContextMany/simple-36                                    12.0k ± 0%     12.0k ± 0%     ~     (all equal)
NewContextMany/WithValuesBefore-36                          12.2k ± 0%     12.2k ± 0%     ~     (all equal)
NewContextMany/WithValuesAfter-36                           12.8k ± 0%     12.8k ± 0%     ~     (all equal)
NewContextMany/WithName-36                                  13.2k ± 0%     13.2k ± 0%     ~     (all equal)
NewContextMany/WithNameAndValuesAfter-36                    14.0k ± 0%     14.0k ± 0%     ~     (all equal)
NewContextMany/WithNameAndValuesBefore-36                   13.4k ± 0%     13.4k ± 0%     ~     (all equal)

```

</details>

Reply:
The logr.go analysis also doesn't look good:
```
< ./logr.go:523:23: make([]interface {}, 0, 64 * len(l.contextKV.handlers)) escapes to heap
---
> ./logr.go:510:35: leaking param: buffer
28a29
> ./logr.go:367:6: moved to heap: buffer
```

- 510: `func (l Logger) valuesFromContext(buffer []interface{}) []interface{} {`
- 367: `var buffer [4]interface{}`

So it's not staying on the stack after all. But why?

Perhaps because it might get accessed by some other Goroutine when the stack of the first one is gone?

But isn't that always the case? I wonder whether the corresponding klog optimization actually works:

https://github.com/kubernetes/klog/blob/d738e55e1450a3219a8c244003042ba290711105/klog.go#L469-L470

Reply:
The difference is that klog uses a global variable to hold the array, not the stack.

By moving the buffer into `contextKV`, I get the expected smaller number of allocations, but making `contextKV` larger negates that improvement.

I consider this a failed idea and will not pursue it further... at least I learned something.

Reply:
ACK